main.js

```diff
@@ -25,31 +24,99 @@ const artifact = require('@actions/artifact');
       charmcraft_channel,
     ]);
 
-    await exec.exec('charmcraft', ['pack', '--destructive-mode', '-q']);
+    chdir(charm_path);
+    core.exportVariable('CHARMCRAFT_AUTH', credentials);
+
+    const metadata = yaml.load(fs.readFileSync('metadata.yaml'));
+
+    const name = metadata.name;
+    const images = Object.entries(metadata.resources)
+      .filter(([_, res]) => res.type === 'oci-image')
+      .map(([name, res]) => [name, res['upstream-source']]);
+
+    const ctx = github.context;
+    const event = ctx.eventName;
+
+    let channel;
+
+    if (event == "push") {
+      if (ctx.ref.startsWith('refs/heads/')) {
+        let branch = ctx.ref.replace('refs/heads/', '');
+
+        if (branch == ctx.payload.repository.master_branch) {
+          channel = 'latest/edge';
+        } else if (branch.startsWith('track/')) {
+          channel = branch.replace('track/', '') + '/edge';
+        } else {
+          core.notice(`Unhandled branch name ${ctx.ref}`);
+          return
+        }
+      } else {
+        core.setFailed(`Unknown type of ref: ${github.context.ref}`);
+        return
+      }
+    } else if (event == "pull_request") {
+      const base_ref = ctx.payload.pull_request.base.ref;
+      const head_ref = ctx.payload.pull_request.head.ref;
+
+      if (head_ref.startsWith('branch/')) {
+        const branch = head_ref.replace('branch/', '');
+
+        if (base_ref == ctx.payload.pull_request.base.repo.default_branch) {
+          channel = `latest/edge/${branch}`;
+        } else if (base_ref.startsWith('track/')) {
+          const track = base_ref.replace('track/', '');
+          channel = `${track}/edge/${branch}`;
+        } else {
+          core.setFailed(`Unhandled PR base name ${base_ref}`);
+          return
+        }
+      } else {
+        core.setFailed(`Unhandled branch name: ${head_ref}`);
+        return
+      }
+    } else {
+      core.setFailed(`Unknown eventType ${event}.`)
+      return;
+    }
+
+    await exec.exec('charmcraft', ['pack', '--destructive-mode', '--quiet']);
```

Comment:
For CI purposes, a more verbose output is desired. The `charmcraft` command is not overwhelmingly verbose without the `-v` flag, I'd remove the `--quiet` flag from this and some other commands below.

Reply:
@DnPlas the `--quiet` flag only gets rid of the progress indication, which IMO is just lots of noise here. The verbose logs are captured by the `finally` block at the bottom that uploads an artifact with all charmcraft logs that it finds. As an example, see these uploaded logs:

https://github.com/canonical/charmhub-upload-action/suites/4581957389/artifacts/123431670

Reply:
I guess the artifacts will be enough.